design/embedded-cluster-cli-driven-update.md

```diff
@@ -0,0 +1,278 @@
+
+**New Embedded Cluster Approach (CLI-First):**
+- Direct CLI commands for all operations without web UI dependency
+- Self-updating binary that can manage both itself and applications
+- All management happens through command-line interface
+- No requirement for in-cluster web console
+
+The implementation will **modify the existing `./binary update` command** to perform binary self-updating when `ENABLE_V3=1` is set, instead of its current behavior of updating applications with airgap bundles.
+
+## Architecture Overview
+
+The self-update mechanism will leverage the existing embedded cluster download infrastructure while adding new components for version management, update detection, and safe binary replacement.
+
+```
+Embedded Cluster Binary → Version Check → Download New Binary → Atomic Replacement
+    ↓                        ↓              ↓                    ↓
+[Current Binary]    [Version API]    [Download Endpoint]    [selfupdate]
+```
+
+## High-Level Changes Required
+
+### 1. Command Line Interface Changes
+
+#### Modified Update Command Behavior
+- **Conditional behavior**: Check `ENABLE_V3` environment variable to determine update mode
+- **V2 mode**: Current behavior (application airgap bundle updates)
+- **V3 mode**: Binary self-updating behavior
+- **Command signature (V3)**: `./binary update --license <license-file> [--version=<version>] [--force]`
+- **License requirement**: The `--license` flag is **required** when `ENABLE_V3=1` for authentication with replicated.app
+- **Validation**: Verify the binary can be updated (not installed via package manager)
+
+#### Command Options (V3 Mode)
+- `--license`: **Required** - License file for authentication with replicated.app
+- `--version`: Optional parameter to specify target version (defaults to latest)
+- `--force`: Skip confirmation prompts for automated scenarios
+
+#### Backward Compatibility
+- **Preserve V2 behavior**: When `ENABLE_V3` is not set or set to 0, maintain current airgap bundle update functionality
+
+### 2. Version Management Infrastructure
+
+**Current State:** Most version infrastructure already exists in the embedded cluster binary.
+
+**What's Already Available:**
+- Version information: `pkg/versions/versions.go` (EC version, k0s version)
+- Embedded release data: App slug, channel, version label via `pkg/release/release.go`
+- Version display: Existing `./binary version` command
+
+#### Architecture Approach
+
+**Dual Sequence Handling:**
+The embedded cluster must handle multiple versioning schemes simultaneously, each serving different purposes:
+
+- **Upstream cursors**: Channel sequence numbers from API (e.g., 1247, 1248, 1251, 1255)
+  - Used for API synchronization and incremental updates
+  - Can have gaps when releases are skipped or removed
+  - Channel-specific and may restart across different channels
+  - Essential for efficient API communication
+
+- **Version labels**: User-facing identifiers (e.g., v1.2.0, v1.2.1)
```

Comment:
why not just support version labels? what's the need to support channel sequences and embedded cluster technical versions?

Reply:
i had the same question, and raised it at the bottom of the doc in the questions section. I think just supported version labels is the way to go as well.